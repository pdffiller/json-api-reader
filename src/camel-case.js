import { mapKeys } from './fn';

const lowerCase = str => str.toLowerCase();
const upperCase = str => str.toUpperCase();

export const toCamelCase = str => str
  .replace(/[\s_-](.)/g, upperCase)
  .replace(/[\s_-]/g, '')
  .replace(/^(.)/, lowerCase);

export const toCamelCaseKeys = mapKeys(toCamelCase);
