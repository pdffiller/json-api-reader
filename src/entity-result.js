import { getAllOrBy } from './fn';

const fillResult = result => entity => {
  result.push(entity.id);
  return entity;
};

export const createResult = (res = []) => ({
  get: getAllOrBy(res),
  fill: fillResult(res),
});
