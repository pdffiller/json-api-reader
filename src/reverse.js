import { compose, map, mapItems, identity, getFrom, mapValues, delProp, addProp } from './fn';
import { toSnakeCase } from './snake-case';


const packSource = source => [source, {}];

const addType = type => ([source, dest]) => [source, {
  ...dest,
  id: source.id,
  type
}];

const getName = (refs, key) => (
  typeof refs[key] === 'string' ? refs[key] : toSnakeCase(key)
);

const toRelation = type => id => ({ type, id });

const buildRelations = (refs, source) => (rels, key) => (
  Boolean(refs[key]) ? Object.assign(rels, {
    [toSnakeCase(key)]: { data: map(source[key], toRelation(getName(refs, key))) },
  }) : rels
);

const __addRefs = (refs, keys) => ([source, dest]) => [source, {
  ...dest,
  relationships: keys.reduce(buildRelations(refs, source), {}),
}];

const _addRefs = refs => {
  const keys = Object.keys(refs);
  return keys.length > 0 ? __addRefs(refs, keys) : identity;
};

const toObj = refs => (
  typeof refs === 'string' ? { [refs]: toSnakeCase(refs) } :
  Array.isArray(refs) ? refs.reduce(
    (nr, key) => Object.assign(nr, { [key]: toSnakeCase(key) }), {}
  ) :
  typeof refs === 'object' ? refs :
  {}
);

const toArray = refs => (
  typeof refs === 'string' ? [refs] :
  Array.isArray(refs) ? refs :
  typeof refs === 'object' ? Object.keys(refs) :
  []
);

const addRefs = compose(_addRefs, toObj);

const readDest = ([, dest]) => dest;

const buildAttrs = (attrs, exclude) => mapItems(
  key => (exclude.includes(key) || !attrs[key] ? null : getName(attrs, key)),
  identity
);

const addAttrs = (attrs, exclude) => ([source, dest]) => [source, {
  ...dest,
  attributes: buildAttrs(toObj(attrs || Object.keys(source)), exclude)(source)
}];

export const jsonApiEntity = ({ type, refs, attrs }) => compose(
  readDest,
  addRefs(refs),
  addAttrs(attrs, toArray(refs).concat('id')),
  addType(toSnakeCase(type)),
  packSource
);

const buildMapper = (mappers, entities, type) => (
  entities && entities[type]
    ? compose(mappers[type], getFrom(entities[type]))
    : mappers[type]
);

const addIncluded = addProp('included', included => included && included.length > 0);

const getIncluded = (mappers, entities, values) => mapValues(
  (items, type) => map(items, buildMapper(mappers, entities, type))
)(values);

export const jsonApiMessage = (mappers, mainType) => (values, entities) => addIncluded(
  {
    data: map(values[mainType], buildMapper(mappers, entities, mainType)),
  },
  getIncluded(mappers, entities, delProp(values, mainType))
);
