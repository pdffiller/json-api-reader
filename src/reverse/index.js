import { jsonApiEntity } from './entity';
import { jsonApiMessage } from './message';

export { jsonApiEntity, jsonApiMessage };

