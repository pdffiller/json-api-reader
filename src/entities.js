import { map } from './fn';
import { mapItems, mapKeys } from './fn-obj';
import { toCamelCase } from './camel-case';

export const $type = Symbol('__$type');
export const $meta = '__$meta';

const rndId = () => Math.random().toFixed(20).slice(2);

const toRelation = (value, key) => ({
  [toCamelCase(key)]: map(value.data, data => data && data.id),
});

export const readEntity = ({ id, type, attributes, relationships, meta }) => ({
  id: id || rndId(),
  [$type]: toCamelCase(type),
  ...(meta && { [$meta]: meta }),
  ...(attributes && mapKeys(attributes, toCamelCase)),
  ...(relationships && mapItems(relationships, toRelation)),
});
