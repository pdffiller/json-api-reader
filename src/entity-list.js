import { $type } from './entities';
import { getAllOrBy } from './fn';

const fillList = entities => entity => {
  const type = entity[$type];
  const e = { ...entity };
  delete e[$type];
  entities[type] = entities[type] || [];
  entities[type].push(e);
  return entity;
};

export const createEntitiesList = (all = {}) => ({
  get: getAllOrBy(all),
  fill: fillList(all),
});
