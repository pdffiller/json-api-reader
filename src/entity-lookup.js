import { $type } from './entities';
import { getAllOrBy } from './fn';

const fillLookup = entities => entity => {
  const type = entity[$type];
  const e = { ...entity };
  delete e[$type];
  entities[type] = entities[type] || {};
  entities[type][entity.id] = e;
  return entity;
};

export const createEntitiesLookup = (all = {}) => ({
  get: getAllOrBy(all),
  fill: fillLookup(all),
});
