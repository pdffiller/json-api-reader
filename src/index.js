import { readEntity } from './entities';
import { forEach, compose } from './fn';
import { createEntitiesLookup } from './entity-lookup';
import { createResult } from './entity-result';
import { createEntitiesList } from './entity-list';

const read = createEntities => ({ data, included }) => {
  const entities = createEntities();
  const result = createResult();

  forEach(data, compose(entities.fill, result.fill, readEntity));
  forEach(included, compose(entities.fill, readEntity));

  return ({
    entities: entities.get(),
    result: result.get(Array.isArray(data) ? null : 0),
  });
};

export const readAsLookup = read(createEntitiesLookup);
export const readAsList = read(createEntitiesList);
