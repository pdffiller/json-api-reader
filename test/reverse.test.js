import { jsonApiEntity, jsonApiMessage } from '../src/reverse';


describe('jsonApiEntity', () => {
  test('should create a mapper function', () => {
    const mapper = jsonApiEntity({
      type: 'Account'
    });
    expect(mapper).toBeInstanceOf(Function);
  });

  test('should map with type only', () => {
    const mapper = jsonApiEntity({
      type: 'Accounts'
    });
    const value = {
      id: '2600-0-0000',
      title: 'Current Account',
    };

    expect(mapper(value)).toMatchSnapshot();
  });

  test('should map with refs', () => {
    const mapper = jsonApiEntity({
      type: 'articles',
      refs: { author: 'people', comments: true },
    });
    const article = require('./__fixtures__/base-result-list').entities.articles[0];
    expect(mapper(article)).toMatchSnapshot();
  });

  test('should map with attrs', () => {
    const mapper = jsonApiEntity({
      type: 'people',
      attrs: ['lastName', 'twitter'],
    });
    const user = require('./__fixtures__/base-result-list').entities.people[0];
    expect(mapper(user)).toMatchSnapshot();
  });
});

describe('jsonApiMessage', () => {
  test('should create a function', () => {
    const Account = jsonApiEntity({ type: 'accounts' });
    const selector = jsonApiMessage({ Account }, 'Account');
    expect(selector).toBeInstanceOf(Function);
  });

  test('should create a valid json-api message', () => {
    const Account = jsonApiEntity({ type: 'accounts' });
    const selector = jsonApiMessage({ Account }, 'Account');
    const values = {
      Account: [
        {
          id: '2600-0-0000',
          title: 'Current Account',
        },
        {
          id: '2605-0-0000',
          title: 'Card Account',
        }
      ]
    };
    expect(
      selector(values)
    ).toMatchSnapshot();
  });

  test('should create a valid message in complex case', () => {
    const { entities, result } = require('./__fixtures__/base-result');
    const articles = jsonApiEntity({
      type: 'articles',
      refs: { author: 'people', comments: true },
    });
    const people = jsonApiEntity({ type: 'people' });
    const comments = jsonApiEntity({
      type: 'comments',
      refs: { author: 'people' }
    });

    const selector = jsonApiMessage({ articles, people, comments }, 'articles');
    const values = {
      articles: result,
      people: Object.keys(entities.people),
      comments: Object.keys(entities.comments),
    };

    const message = selector(values, entities);

    expect(message).toMatchSnapshot();
    expect(
      selector({
        ...values,
        articles: result[0]
      }, entities)
    ).toMatchSnapshot();
  });

  test('should create a valid message in complex case', () => {
    const { entities: values } = require('./__fixtures__/base-result-list');
    const articles = jsonApiEntity({
      type: 'articles',
      refs: { author: 'people', comments: true },
    });
    const people = jsonApiEntity({ type: 'people' });
    const comments = jsonApiEntity({
      type: 'comments',
      refs: { author: 'people' }
    });

    const selector = jsonApiMessage({ articles, people, comments }, 'articles');

    const message = selector(values);

    expect(message).toMatchSnapshot();
  });
});
