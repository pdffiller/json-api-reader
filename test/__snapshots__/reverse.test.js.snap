// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonApiEntity should map with attrs 1`] = `
Object {
  "attributes": Object {
    "last_name": "Gebhardt",
    "twitter": "dgeb",
  },
  "id": "9",
  "type": "people",
}
`;

exports[`jsonApiEntity should map with refs 1`] = `
Object {
  "attributes": Object {
    "title": "JSON API paints my bikeshed!",
  },
  "id": "1",
  "relationships": Object {
    "author": Object {
      "data": Object {
        "id": "9",
        "type": "people",
      },
    },
    "comments": Object {
      "data": Array [
        Object {
          "id": "5",
          "type": "comments",
        },
        Object {
          "id": "12",
          "type": "comments",
        },
      ],
    },
  },
  "type": "articles",
}
`;

exports[`jsonApiEntity should map with type only 1`] = `
Object {
  "attributes": Object {
    "title": "Current Account",
  },
  "id": "2600-0-0000",
  "type": "accounts",
}
`;

exports[`jsonApiMessage should create a valid json-api message 1`] = `
Object {
  "data": Array [
    Object {
      "attributes": Object {
        "title": "Current Account",
      },
      "id": "2600-0-0000",
      "type": "accounts",
    },
    Object {
      "attributes": Object {
        "title": "Card Account",
      },
      "id": "2605-0-0000",
      "type": "accounts",
    },
  ],
}
`;

exports[`jsonApiMessage should create a valid message in complex case 1`] = `
Object {
  "data": Array [
    Object {
      "attributes": Object {
        "title": "JSON API paints my bikeshed!",
      },
      "id": "1",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
        "comments": Object {
          "data": Array [
            Object {
              "id": "5",
              "type": "comments",
            },
            Object {
              "id": "12",
              "type": "comments",
            },
          ],
        },
      },
      "type": "articles",
    },
  ],
  "included": Array [
    Object {
      "attributes": Object {
        "first_name": "Dan",
        "last_name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "type": "people",
    },
    Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "2",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
    Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  ],
}
`;

exports[`jsonApiMessage should create a valid message in complex case 2`] = `
Object {
  "data": Object {
    "attributes": Object {
      "title": "JSON API paints my bikeshed!",
    },
    "id": "1",
    "relationships": Object {
      "author": Object {
        "data": Object {
          "id": "9",
          "type": "people",
        },
      },
      "comments": Object {
        "data": Array [
          Object {
            "id": "5",
            "type": "comments",
          },
          Object {
            "id": "12",
            "type": "comments",
          },
        ],
      },
    },
    "type": "articles",
  },
  "included": Array [
    Object {
      "attributes": Object {
        "first_name": "Dan",
        "last_name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "type": "people",
    },
    Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "2",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
    Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  ],
}
`;

exports[`jsonApiMessage should create a valid message in complex case 3`] = `
Object {
  "data": Array [
    Object {
      "attributes": Object {
        "title": "JSON API paints my bikeshed!",
      },
      "id": "1",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
        "comments": Object {
          "data": Array [
            Object {
              "id": "5",
              "type": "comments",
            },
            Object {
              "id": "12",
              "type": "comments",
            },
          ],
        },
      },
      "type": "articles",
    },
  ],
  "included": Array [
    Object {
      "attributes": Object {
        "first_name": "Dan",
        "last_name": "Gebhardt",
        "twitter": "dgeb",
      },
      "id": "9",
      "type": "people",
    },
    Object {
      "attributes": Object {
        "body": "First!",
      },
      "id": "5",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "2",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
    Object {
      "attributes": Object {
        "body": "I like XML better",
      },
      "id": "12",
      "relationships": Object {
        "author": Object {
          "data": Object {
            "id": "9",
            "type": "people",
          },
        },
      },
      "type": "comments",
    },
  ],
}
`;
